crates/sui-verifier/src/private_generics.rs

```diff
@@ -0,0 +1,185 @@
+    // do not need to check the sui::transfer module itself
+    for func_def in &module.function_defs {
+        verify_function(view, func_def).map_err(|error| {
+            verification_failure(format!(
+                "{}::{}. {}",
+                module.self_id(),
+                module.identifier_at(module.function_handle_at(func_def.function).name),
+                error
+            ))
+        })?;
+    }
+    Ok(())
+}
+
+fn verify_function(view: &BinaryIndexedView, fdef: &FunctionDefinition) -> Result<(), String> {
+    let code = match &fdef.code {
+        None => return Ok(()),
+        Some(code) => code,
+    };
+    let function_type_parameters = &view.function_handle_at(fdef.function).type_parameters;
+    for instr in &code.code {
+        if let Bytecode::CallGeneric(finst_idx) = instr {
+            let FunctionInstantiation {
+                handle,
+                type_parameters,
+            } = view.function_instantiation_at(*finst_idx);
+
+            let fhandle = view.function_handle_at(*handle);
+            let mhandle = view.module_handle_at(fhandle.module);
+
+            let type_arguments = &view.signature_at(*type_parameters).0;
+            match addr_module(view, mhandle) {
+                (SUI_FRAMEWORK_ADDRESS, "transfer") => verify_private_transfer(
+                    view,
+                    function_type_parameters,
+                    fhandle,
+                    type_arguments,
+                )?,
+                (SUI_FRAMEWORK_ADDRESS, "event") => {
+                    verify_private_event_emit(view, fhandle, type_arguments)?
+                }
+                _ => (),
+            }
+        }
+    }
+    Ok(())
+}
+
+fn verify_private_transfer(
+    view: &BinaryIndexedView,
+    function_type_parameters: &[AbilitySet],
+    fhandle: &FunctionHandle,
+    type_arguments: &[SignatureToken],
+) -> Result<(), String> {
+    let self_handle = view.module_handle_at(view.self_handle_idx().unwrap());
+    if addr_module(view, self_handle) == (SUI_FRAMEWORK_ADDRESS, "transfer") {
+        return Ok(());
+    }
+    let fident = view.identifier_at(fhandle.name);
+    match fident.as_str() {
+        // transfer functions
+        "transfer"
+        | "transfer_to_object_id"
+        | "freeze_object"
+        | "share_object"
+        | "transfer_to_object"
+        | "transfer_to_object_unsafe"
+        | "transfer_child_to_object"
+        | "transfer_child_to_address" => (),
+        // these functions operate over ChildRef
+        "is_child" | "is_child_unsafe" | "delete_child_object" => {
+            return Ok(());
+        }
+        // should be unreachable
+        // these are private and the module itself is skipped
+        "transfer_internal" | "delete_child_object_internal" => {
+            debug_assert!(false);
+            return Ok(());
+        }
+        // unknown function, so a bug in the implementation here
+        s => {
```

Comment:
Love the exhaustiveness check! I predict this will catch a bug in the future :).